File: sea-shell-service-spring-web-flux/sea-shell-rest-service/src/main/java/org/jesperancinha/shell/webflux/service/SeaShellReactiveOneServiceImpl.java
package org.jesperancinha.shell.webflux.service;

import lombok.extern.slf4j.Slf4j;
import org.jesperancinha.shell.webflux.data.SeaShellAccountDto;
import org.jesperancinha.shell.webflux.data.SeaShellCostumeDto;
import org.jesperancinha.shell.webflux.data.SeaShellDto;
import org.jesperancinha.shell.webflux.data.SeaShellLowerDto;
import org.jesperancinha.shell.webflux.data.SeaShellPersonDto;
import org.jesperancinha.shell.webflux.data.SeaShellTopDto;
import org.jesperancinha.shell.webflux.repo.ShellAccountRepository;
import org.jesperancinha.shell.webflux.repo.ShellCostumeRepository;
import org.jesperancinha.shell.webflux.repo.ShellLowerRepository;
import org.jesperancinha.shell.webflux.repo.ShellPersonRepository;
import org.jesperancinha.shell.webflux.repo.ShellRepository;
import org.jesperancinha.shell.webflux.repo.ShellTopRepository;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Slf4j
@Service
public class SeaShellReactiveOneServiceImpl extends SeaShellOneAdapter implements SeaShellReactiveOneService {


    public SeaShellReactiveOneServiceImpl(ShellRepository shellRepository,
                                          ShellPersonRepository shellPersonRepository,
                                          ShellCostumeRepository shellCostumeRepository,
                                          ShellAccountRepository shellAccountRepository,
                                          ShellTopRepository shellTopRepository,
                                          ShellLowerRepository shellLowerRepository) {
        super(
                shellRepository,
                shellPersonRepository,
                shellCostumeRepository,
                shellAccountRepository,
                shellTopRepository,
                shellLowerRepository);
    }

    public Flux<Long> getAllIds() {
        return this.shellRepository.findAllShellIds();
    }

    public Mono<SeaShellDto> getSeaShellById(Long id) {
        return this.shellRepository.findSeaShellById(id)
                .map(SeaShellConverter::toShellDto);
    }

    public Mono<SeaShellPersonDto> getPersonById(Long id) {
        return this.shellPersonRepository.findPersonById(id)
                .map(SeaShellConverter::toShellPersonDto);
    }

    public Mono<SeaShellCostumeDto> getCostumeById(Long id) {
        return this.shellCostumeRepository.findCostumeById(id)
                .map(SeaShellConverter::toShellCostumeDto);
    }

    public Mono<SeaShellAccountDto> getAccountById(String id) {
        return this.shellAccountRepository.findAccountById(id)
                .map(SeaShellConverter::toAccountDto);
    }

    public Mono<SeaShellTopDto> getTopById(Long id) {
        return this.shellTopRepository.findTopById(id)
                .map(SeaShellConverter::toTopDto);
    }

    public Mono<SeaShellLowerDto> getLowerById(Long id) {
        return this.shellLowerRepository.findLowerById(id)
                .map(SeaShellConverter::toLowerDto);

    }
}

File: sea-shell-service-spring-web-flux/src/main/kotlin/org/jesperancinha/shell/webflux/service/SeaShellReactiveOneServiceImpl.kt
package org.jesperancinha.shell.webflux.service

import lombok.extern.slf4j.Slf4j
import org.jesperancinha.shell.client.accounts.Account
import org.jesperancinha.shell.client.costumes.Costume
import org.jesperancinha.shell.client.lowers.Lower
import org.jesperancinha.shell.client.persons.Person
import org.jesperancinha.shell.client.shells.Shell
import org.jesperancinha.shell.client.tops.Top
import org.jesperancinha.shell.webflux.data.*
import org.jesperancinha.shell.webflux.repo.*
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

@Slf4j
@Service
class SeaShellReactiveOneServiceImpl(
    shellRepository: ShellRepositoryImpl,
    shellPersonRepository: ShellPersonRepositoryImpl,
    shellCostumeRepository: ShellCostumeRepositoryImpl,
    shellAccountRepository: ShellAccountRepositoryImpl,
    shellTopRepository: ShellTopRepositoryImpl,
    shellLowerRepository: ShellLowerRepositoryImpl
) : SeaShellOneAdapter(
    shellRepository,
    shellPersonRepository,
    shellCostumeRepository,
    shellAccountRepository,
    shellTopRepository,
    shellLowerRepository
) {
    fun allIds(): Flux<Long> = shellRepository.findAllShellIds()

    fun getSeaShellById(id: Long): Mono<SeaShellDto> {
        return shellRepository.findSeaShellById(id)
            .map { SeaShellConverter.toShellDto(it) }
    }

    fun getPersonById(id: Long): Mono<SeaShellPersonDto> {
        return shellPersonRepository.findPersonById(id)
            .map { SeaShellConverter.toShellPersonDto(it) }
    }

    fun getCostumeById(id: Long): Mono<SeaShellCostumeDto> {
        return shellCostumeRepository.findCostumeById(id)
            .map { SeaShellConverter.toShellCostumeDto(it) }
    }

    fun getAccountById(id: String): Mono<SeaShellAccountDto> {
        return shellAccountRepository.findAccountById(id)
            .map { SeaShellConverter.toAccountDto(it) }
    }

    fun getTopById(id: Long): Mono<SeaShellTopDto> {
        return shellTopRepository.findTopById(id)
            .map { SeaShellConverter.toTopDto(it) }
    }

    fun getLowerById(id: Long): Mono<SeaShellLowerDto> {
        return shellLowerRepository.findLowerById(id)
            .map { SeaShellConverter.toLowerDto(it) }
    }
}